analyzer/src/main/kotlin/managers/utils/DependencyGraphBuilder.kt

```diff
@@ -0,0 +1,329 @@
+ * this is referred to as "fragmentation": A fragment is a consistent sub graph, in which each package occurs only
+ * once. Packages appearing multiple times with different dependencies need to be placed in separate fragments. It is
+ * then possible to uniquely identify a specific package by a combination of its numeric identifier and the index of
+ * the fragment it belongs to.
+ *
+ * This class implements the full logic to construct a [DependencyGraph], independent on the concrete representation of
+ * dependencies used by specific package managers. To make this class compatible with such a dependency representation,
+ * the package manager implementation has to provide a [DependencyHandler]. Via this handler, all the relevant
+ * information about dependencies can be extracted.
+ */
+class DependencyGraphBuilder<D>(
+    /**
+     * The [DependencyHandler] used by this builder instance to extract information from the dependency objects when
+     * constructing the [DependencyGraph].
+     */
+    val dependencyHandler: DependencyHandler<D>
+) {
+    /**
+     * A list storing the identifiers of all dependencies added to this builder. This list is then used to resolve
+     * dependencies based on their indices.
+     */
+    private val dependencyIds = mutableListOf<String>()
+
+    /**
+     * A mapping of the identifiers of the dependencies known to this builder to their numeric indices.
+     */
+    private val dependencyIndexMapping = mutableMapOf<String, Int>()
+
+    /**
+     * Stores all the references to dependencies that have been added so far. Each element of the list represents a
+     * fragment of the dependency graph. For each fragment, there is a mapping from a dependency index to the
+     * reference pointing to the corresponding dependency tree.
+     */
+    private val referenceMappings = mutableListOf<MutableMap<Int, DependencyReference>>()
+
+    /** The mapping from scopes to dependencies constructed by this builder. */
+    private val scopeMapping = mutableMapOf<String, List<RootDependencyIndex>>()
+
+    /** Stores all packages encountered in the dependency tree. */
+    private val resolvedPackages = mutableSetOf<Package>()
+
+    /**
+     * A set storing the packages that are direct dependencies of one of the scopes. These are the entry points into
+     * the dependency graph.
+     */
+    private val directDependencies = mutableSetOf<DependencyReference>()
+
+    /**
+     * Add the scope with the given [scopeName] to this builder. In most cases, it is not necessary to add scopes
+     * explicitly, as they are recorded automatically by _addDependency()_. However, if there are scopes without
+     * dependencies, this function can be used to include them into the builder result.
+     */
+    fun addScope(scopeName: String) {
+        scopeMapping.putIfAbsent(scopeName, emptyList())
+    }
```

Comment:
Nit: Could be a function expression (if returning the return value of `putIfAbsent` is fine with you).

Reply:
I think, I would rather stick to the return type `Unit` here. Otherwise, the return type would be a nullable list of `RootDependencyIndex`, which does not make much sense for the caller. As the return type is part of the public API, it is IMHO better to explicitly state, that nothing meaningful is returned.